From d24384d091eec141d8d2e853bdf8ab9cb75ce069 Mon Sep 17 00:00:00 2001
From: Eric Anholt <eric@anholt.net>
Date: Tue, 7 Mar 2017 17:21:25 +1100
Subject: [PATCH 13/36] mesa: Connect the generated GL command marshalling code
 to the build.

v2: Rebase on the Begin/End changes, and just disable this feature on
    non-GL-core.
v3: (Timothy Arceri) enable for non-GL-core contexts. Remove
    unrelated safe_mul() hunk. while loop style fix.
---
 src/mesa/Makefile.sources |  2 ++
 src/mesa/main/glthread.c  | 15 +++++++++++++++
 2 files changed, 17 insertions(+)

diff --git a/src/mesa/Makefile.sources b/src/mesa/Makefile.sources
index b2d078368a..134d5e94f5 100644
--- a/src/mesa/Makefile.sources
+++ b/src/mesa/Makefile.sources
@@ -132,6 +132,8 @@ MAIN_FILES = \
 	main/lines.c \
 	main/lines.h \
 	main/macros.h \
+	main/marshal_generated.c \
+	main/marshal_generated.h \
 	main/matrix.c \
 	main/matrix.h \
 	main/mipmap.c \
diff --git a/src/mesa/main/glthread.c b/src/mesa/main/glthread.c
index 8877a695f0..c4d1031f4c 100644
--- a/src/mesa/main/glthread.c
+++ b/src/mesa/main/glthread.c
@@ -54,8 +54,15 @@ glthread_allocate_batch(struct gl_context *ctx)
 static void
 glthread_unmarshal_batch(struct gl_context *ctx, struct glthread_batch *batch)
 {
+   size_t pos = 0;
+
    _glapi_set_dispatch(ctx->CurrentServerDispatch);
 
+   while (pos < batch->used)
+      pos += _mesa_unmarshal_dispatch_cmd(ctx, &batch->buffer[pos]);
+
+   assert(pos == batch->used);
+
    free(batch->buffer);
    free(batch);
 }
@@ -112,6 +119,14 @@ _mesa_glthread_init(struct gl_context *ctx)
    if (!glthread)
       return;
 
+   ctx->MarshalExec = _mesa_create_marshal_table(ctx);
+   if (!ctx->MarshalExec) {
+      free(glthread);
+      return;
+   }
+
+   ctx->CurrentClientDispatch = ctx->MarshalExec;
+
    pthread_mutex_init(&glthread->mutex, NULL);
    pthread_cond_init(&glthread->new_work, NULL);
    pthread_cond_init(&glthread->work_done, NULL);
-- 
2.12.0

